userdetails-plugin/grails-app/views/registration/createAccount.gsp

```diff
@@ -422,6 +446,110 @@
 
+         $("#requestCode").click(function(e) {
+             var newEmail = $("#newEmail").val();
+             if(newEmail == null || newEmail === "") {
+                 document.getElementById("newEmailMessage").innerHTML = "Invalid email"
+                 document.getElementById("newEmailMessage").style.color = "red"
+                 document.getElementById("newEmailMessage").hidden = false
+             }
+             else {
+                 $.ajax({
+                 url: "${createLink(uri:'/registration/update')}?email=" + newEmail,
+                 type: "GET",
+                 success: function(result){
+                     let error = "A user is already registered with the email address"
+                     if(result.success) {
+                        document.getElementById("newEmailMessage").innerHTML = "Please enter the code received in your new email"
+                        document.getElementById("newEmailMessage").style.color = "green"
+                        document.getElementById("newEmailMessage").hidden = false
+                        document.getElementById("emailCodeDiv").hidden = false
+                     }
+                     else if(result.includes("Failed to update user profile")){
+                         if(result.includes(error)){
+                             document.getElementById("newEmailMessage").innerHTML = error
+                         }
+                         else {
+                            document.getElementById("newEmailMessage").innerHTML = 'Error please contact us at <a href="mailto:support@ala.org.au">support@ala.org.au</a>'
+                         }
+                         document.getElementById("newEmailMessage").style.color = "red"
+                         document.getElementById("newEmailMessage").hidden = false
+                         document.getElementById("emailCodeDiv").hidden = true
+                    }
+                }});
+             }
+         });
+
+         $("#verifyCode").click(function(e) {
+            var emailCode = $("#emailCode").val();
+             if(emailCode == null || emailCode === "") {
+                 document.getElementById("newEmailMessage").innerHTML = "Invalid code"
+                 document.getElementById("newEmailMessage").style.color = "red"
+                 document.getElementById("newEmailMessage").hidden = false
+                 document.getElementById("emailCodeDiv").hidden = false
+             }
+             else {
+                 $.ajax({
+                 url: "${createLink(uri:'/registration/verifyAttributeChangeWithCode')}?attribute=email&code=" + emailCode,
+                 type: "GET",
+                 success: function(result){
+                     if(result.success){
+                        window.location = "${createLink(uri:'/logout')}?url=/"
+                    }
+                     else{
+                         document.getElementById("newEmailMessage").innerHTML = result.error
```

Comment:
Potential XSS issue? If the userservice exception reflects the user input code or something then this is a problem.

Reply:
This is why I prefer to never touch innerHTML from jquery.